massa-execution/src/worker.rs

```diff
@@ -255,106 +253,137 @@ impl ExecutionWorker {
+                            // no subsequent CSS-final block found in the same thread as cur_slot
+                            break;
+                        }
+                    }
 
-        // list SCE-final slots/blocks
-        for (b_id, block) in css_final_blocks.into_iter() {
-            let block_slot = block.header.content.slot;
-            if block_slot <= self.last_final_slot {
-                continue;
-            }
-            loop {
-                let next_final_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
-                if next_final_slot == block_slot {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                    break;
-                } else if next_final_slot < max_thread_slot[next_final_slot.thread as usize] {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                } else {
-                    self.ordered_pending_css_final_blocks.push((b_id, block));
-                    break;
+                    // there are no more CSS-final blocks
+                    None => break,
                 }
+
+                cur_slot = cur_slot.get_next_slot(self.cfg.thread_count)?;
             }
         }
 
-        // list remaining CSS finals + new blockclique
-        self.ordered_active_blocks = self
-            .ordered_pending_css_final_blocks
+        // process SCE-active blocks
+        let mut sce_active_blocks: BTreeMap<Slot, (&BlockId, &Block)> = blockclique
             .iter()
-            .cloned()
+            .filter_map(|(b_id, b)| {
+                if b.header.content.slot <= self.last_final_slot {
+                    // eliminate blocks that are not from a stricly later slot than the current latest SCE-final one
+                    return None;
+                }
+                Some((b.header.content.slot, (b_id, b)))
+            })
             .chain(
-                blockclique
-                    .into_iter()
-                    .filter(|(_b_id, b)| b.header.content.slot > self.last_final_slot),
+                self.pending_css_final_blocks
+                    .iter()
+                    .map(|(k, (b_id, b))| (*k, (b_id, b))),
             )
             .collect();
+        if let Some(max_css_active_slot) = sce_active_blocks.last_key_value().map(|(s, _v)| *s) {
+            let mut cur_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
+            while cur_slot <= max_css_active_slot {
+                match sce_active_blocks.first_key_value().map(|(s, _v)| *s) {
```

Comment:
Please assign `sce_active_blocks.first_key_value().map(|(s, _v)| *s)` to a named variable for clarity.

Reply:
done